fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreBase.java

```diff
@@ -976,12 +1000,80 @@ default RecordCursor<FDBIndexedRecord<M>> scanIndexRecords(@Nonnull final String
+                                                               @Nonnull ScanProperties scanProperties) {
+        int commonPrimaryKeyLength = 0;
+        if (fetchMethod != IndexFetchMethod.SCAN_AND_FETCH) {
+            commonPrimaryKeyLength = getCommonPrimaryKeyLength(index);
+        }
+        return scanIndexRecords(index, fetchMethod, scanBounds, commonPrimaryKeyLength, continuation, orphanBehavior, scanProperties);
+    }
+
+    /**
+     * Scan the records pointed to by an index. This version of the method can choose between various scan options as
+     * determined by the {@link IndexFetchMethod} parameter:
+     * <UL>
+     *     <LI>SCAN_AND_FETCH: Scan the index and fetch each record individually</LI>
+     *     <LI>USE_REMOTE_FETCH: Use FDB's remote fetch feature to scan the index and fetch the records in one call</LI>
+     *     <LI>USE_REMOTE_FETCH_WITH_FALLBACK: Use FDB's remote fetch feature but fall back to SCAN_AND_FETCH in case of failure</LI>
+     * </UL>
+     * This method can only be used with {@link IndexScanRange} scan bounds.
+     * This method requires a positive common primary key length for USE_REMOTE_FETCH and USE_REMOTE_FETCH_WITH_FALLBACK.
+     * Common primary key length is the length (# of elements) that is the same for all types defined for the index.
+     *
+     * @param index the index to scan
+     * @param fetchMethod the fetch method to use when getting records from the index
+     * @param scanBounds the range of the index to scan, has to be a IndexScanRange
+     * @param commonPrimaryKeyLength the length of the common primary key for the records that would be returned, only required for USE_REMOTE_FETCH and USE_REMOTE_FETCH_WITH_FALLBACK
+     * @param continuation any continuation from a previous scan
+     * @param orphanBehavior how the iteration process should respond in the face of entries in the index for which there is no associated record
+     * @param scanProperties skip, limit and other scan properties
+     * @return a cursor that return records pointed to by the index
+     */
+    @API(API.Status.EXPERIMENTAL)
+    @Nonnull
+    default RecordCursor<FDBIndexedRecord<M>> scanIndexRecords(@Nonnull final Index index,
+                                                               @Nonnull final IndexFetchMethod fetchMethod,
+                                                               @Nonnull final IndexScanBounds scanBounds,
+                                                               int commonPrimaryKeyLength,
+                                                               @Nullable byte[] continuation,
+                                                               @Nonnull IndexOrphanBehavior orphanBehavior,
+                                                               @Nonnull ScanProperties scanProperties) {
+        if ((fetchMethod != IndexFetchMethod.SCAN_AND_FETCH) && (commonPrimaryKeyLength <= 0)) {
```

Comment:
Is there a reason that this doesn't remove the `commonPrimaryKeyLength` parameter from here, even, and instead  calculate the `commonPrimaryKeyLength` as the first thing in this function? Arguably, it could be pushed even further down into `scanIndexRemoteFetch`, if we wanted to, I suppose.

Reply:
This is what the signature at line 1031 does. The reason for exposing both the "with" length and the "without" length methods is so that in the cases where the PK length is known (pre calculated) it can be passed in and not re-calculated.